### PR TITLE
Update

### PR DESCRIPTION
**Pull request by hildjj**
Update dependencies.  Update rules.  This will cause a bunch of things to lint-fail in peggy, but it's time for that.

### CODE REVIEW

**Comment by hildjj** [comment]
I'm not sure why I had a pending change to update to 3.0.  It doesn't look like I released it.  I'm going to leave it in, and make this PR be the bases for 3.1, hoping for the best.

**Comment by hildjj** [comment]
@Mingun do you have any opinions about the things I changed?  I'm sorry I didn't do the patch in two steps to make it easier to separate the moves from the changes.

File: index.js [hunk 1]
```diff
@@ -226,33 +307,96 @@ module.exports = {
...
+    "no-restricted-imports": "off",
     "no-restricted-properties": "off",
+    "no-restricted-syntax": "off",
     "no-return-assign": "error",
     "no-return-await": "error",
     "no-script-url": "error",
-    "no-self-assign": "error",
-    "no-self-compare": "error",
     "no-sequences": "error",
+    "no-shadow": "error",
```
Comment (hildjj) [inline index.js:327]: This is going to cause good changes in Peggy.

File: index.js [hunk 2]
```diff
@@ -173,50 +224,80 @@ module.exports = {
...
+    "no-lonely-if": "off",
     "no-loop-func": "error",
 
     // Disabled because sometimes putting a number inline is better than
     // complicating the code by creating a constant.
     "no-magic-numbers": "off",
-
-    // Disabled because multiple spaces are often used for alignment.
-    "no-multi-spaces": "off",
+    "no-mixed-operators": "error",
```
Comment (hildjj) [inline index.js:289]: This is going to cause a few sets of parens, which is worth it, imo.

File: index.js [hunk 3]
```diff
@@ -173,50 +224,80 @@ module.exports = {
...
-    "no-empty-pattern": "off",
+    "no-empty-static-block": "error",
 
     // Enabled, however this isn't strictly necessary because `==` and `!=`
     // are not allowed at all by `eqeqeq`. But it can't hurt.
     "no-eq-null": "error",
-
-    // Disabled because `eval` has legitimate uses.
-    "no-eval": "off",
+    "no-eval": "error",
```
Comment (hildjj) [inline index.js:234]: We'll put an eslint-ignore in the two places where we want eval explicitly.

File: index.js [hunk 4]
```diff
@@ -80,71 +83,119 @@ module.exports = {
...
+    "no-array-constructor": "error",
+
+    // Disabled because bitwise operators have legitimate uses.
+    "no-bitwise": "off",
     "no-caller": "error",
     "no-case-declarations": "error",
-    "no-constructor-return": "error",
-    "no-constant-binary-expression": "error",
-    "no-div-regex": "off",
+    "no-confusing-arrow": "error",
```
Comment (hildjj) [inline index.js:186]: There are a few places that are legitimately confusing.  We'll fix those.

File: ts.js [hunk 5]
```diff
@@ -87,6 +89,7 @@ module.exports = {
     "@typescript-eslint/func-call-spacing": "error",
     "@typescript-eslint/indent": ["off", 2], // Broken, see https://github.com/typescript-eslint/typescript-eslint/issues/1824
     "@typescript-eslint/init-declarations": "error",
```
Comment (hildjj) [inline ts.js:91]: This is going to be a biggish change, but I think it's good for quality.

File: index.js [hunk 6]
```diff
@@ -80,71 +83,119 @@ module.exports = {
...
     "curly": "error",
     "default-case": "error",
     "default-case-last": "error",
     "default-param-last": "error",
-    "dot-location": ["error", "property"],
     "dot-notation": "error",
     "eqeqeq": "error",
+    "func-name-matching": "error",
+    "func-names": "off", // Tedious
+    "func-style": ["error", "declaration"],
```
Comment (hildjj) [inline index.js:149]: This only causes one change in tests.

File: index.js [hunk 7]
```diff
@@ -80,71 +83,119 @@ module.exports = {
...
-    // an interface.
-    "class-methods-use-this": "off",
+
+    // We have underscores in our public API already
+    "camelcase": "off",
+    "capitalized-comments": ["error", "always", {
+      ignorePattern: "c8",
+      ignoreConsecutiveComments: true,
+    }],
+    "class-methods-use-this": "error",
```
Comment (hildjj) [inline index.js:135]: Causes a few things that will have to be ignored (abstract methods) and a few places that we might want to move to static.

File: index.js [hunk 8]
```diff
@@ -80,71 +83,119 @@ module.exports = {
...
-    "no-unreachable-loop": "off",
+    "no-unreachable-loop": "error",
     "no-unsafe-finally": "error",
     "no-unsafe-negation": "error",
     "no-unsafe-optional-chaining": "error",
+    "no-unused-private-class-members": "off", // Not using private yet
+    "no-unused-vars": "error",
+
+    // This is going to cause problems, but I don't care.
+    "no-use-before-define": "error",
```
Comment (hildjj) [inline index.js:109]: This causes a pretty large set of changes, most of which I really like.

**Comment by Mingun** [comment]
It would be easier to me see the changes if you show what changes it led to in peggy (making PR there).

**Comment by hildjj** [comment]
See https://github.com/peggyjs/peggy/pull/349

**Comment by hildjj** [comment]
I didn't mind having to explicitly opt out when I was using practices that were questionable.

File: index.js (hunk 1 above)
Comment (Mingun) [reply index.js:327]: I'm not sure... It creates a lot of strange renames (when function parameter shadows external variable) without much benefit, IMO. But I'm not strongly against it. Maybe enable shadowing for some names (`node`, `context`)?

**Review by Mingun — APPROVED** [review]

File: index.js (hunk 1 above)
Comment (hildjj) [reply index.js:327]: I'll turn it off for now.  Perhaps we can enable it some other time.

**Pull request merged**